### PR TITLE
More iterations

### PR DESCRIPTION
**Pull request by devoncarew**
- run more iterations of the analyzer benchmarks (run 3 iterations and take the best time of the three)
- shorter stack traces when exceptions are thrown through main()
- push the current flutter commit up to firebase w/ the rest of the data
- fix an issue w/ getting the dart sdk version
- fix an exception that occurred with piping streams:
```
Bad state: StreamSink is already bound to a stream
dart:io/io_sink.dart 159                                      _StreamSinkImpl.addStream
dart:io/stdio.dart 265                                        _StdSink.addStream
package:dashboard_box/src/utils.dart 83:10                    exec.<async>
```

@yjbanov @sethladd

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,17 +26,17 @@ On the actual build box `launchd` writes standard output in
...
    upload metrics (not needed for local testing)
 
 Example:
 
 ```json
 {
-  "ANDROID_DEVICE_ID": "...",
-  "FIREBASE_FLUTTER_DASHBOARD_TOKEN": "..."
+  "android_device_id": "...",
+  "firebase_flutter_dashboard_token": "..."
```
Comment (yjbanov) [inline README.md]: Thanks for cleaning it up!

File: lib/src/utils.dart [hunk 2]
```diff
@@ -62,18 +62,32 @@ void section(String title) {
 }
 
 Future<String> getDartVersion() async {
-  String version = await eval(dartBin, ['--version']);
+  // The Dart SDK return the version text to stderr.
+  ProcessResult result = Process.runSync('dart', ['--version']);
```
Comment (yjbanov) [inline lib/src/utils.dart]: Use `dartBin`. `'dart'` points to the version installed from http://dartlang.org.

File: lib/src/utils.dart [hunk 3]
```diff
@@ -62,18 +62,32 @@ void section(String title) {
...
-  stdout.addStream(proc.stdout);
-  stderr.addStream(proc.stderr);
+  proc.stdout
+    .transform(UTF8.decoder)
+    .transform(const LineSplitter())
+    .listen(print);
+  proc.stderr
+    .transform(UTF8.decoder)
+    .transform(const LineSplitter())
+    .listen(stderr.writeln);
```
Comment (yjbanov) [inline lib/src/utils.dart]: Curious, what's the difference?

Comment (devoncarew) [inline lib/src/utils.dart]: W/ the original code, I see:

```
Bad state: StreamSink is already bound to a stream
dart:io/io_sink.dart 159                                      _StreamSinkImpl.addStream
dart:io/stdio.dart 265                                        _StdSink.addStream
package:dashboard_box/src/utils.dart 83:10                    exec.<async>
```

I didn't really suss out why that was happening, but the above works equally well (if a bit more verbose).

File: bin/build.dart [hunk 4]
```diff
@@ -119,44 +127,93 @@ Future<int> runStartupTest(String testDirectory, String testName) {
...
-  });
+  Benchmark benchmark = new FlutterAnalyzeBenchmark(now, sdk, commit);
+  section(benchmark.name);
+  await runBenchmark(benchmark, iterations: 3);
 
-  _patchupAnalysisResult(benchmark, now, expected: 25.0, sdk: sdk);
-  copy(benchmark, config.dataDirectory, name: 'analyzer_cli__analysis_time.json');
+  benchmark = new FlutterAnalyzeAppBenchmark(now, sdk, commit);
+  section(benchmark.name);
+  await runBenchmark(benchmark, iterations: 3);
```
Comment (yjbanov) [inline bin/build.dart]: Tripling iteration count will add 2 more minutes to the run, based on the latest dashboard numbers for the analysis benchmark. There's also the gallery app benchmark that about to land, which is also 2 minutes (if we're lucky). Perhaps we should split the analyzer benchmark into its own script and run it on a separate machine?

Comment (devoncarew) [inline bin/build.dart]: I'm not super concerned with the amount of time the script currently takes to run - it seems reasonable (~10 mins?). I'm much more concerned with getting good (reliable, repeatable) analysis numbers. Once the analysis benchmarks are a little faster, I'll want to bump the iteration count up to 4 or 5x :)

There are other places in the script we could make faster, like the initial startup / setup. Moving to parallelize the script across multiple machines sounds reasonable. Myself, I would wait to do that until we're closer to maxing out what the machine can do in 24 hours. But happy to discuss and help with that if you think it's necessary.

File: lib/src/utils.dart (hunk 2 above)
Comment (devoncarew) [inline lib/src/utils.dart]: updated to `Process.runSync(dartBin, ['--version'])`

File: bin/build.dart [hunk 5]
```diff
@@ -174,3 +231,38 @@ Future<Null> uploadDataToFirebase() async {
...
+
+  while (iterations > 0) {
+    iterations--;
+
+    print('');
+
+    num result = await benchmark.run();
+
+    if (minValue == null || result < minValue) {
+      benchmark.markLastRunWasBest(result);
```
Comment (yjbanov) [inline bin/build.dart]: Why not average + sigma?

File: bin/build.dart (hunk 4 above)
Comment (yjbanov) [inline bin/build.dart]: Sure, let's add it, and chat about splitting into multiple machines later (should be easy now that everything goes to Firebase). To bootstrap the compartmentalization process though, let's move the analysis benchmarking code into its own library under `lib/` (I'll do the same for other benchmarks). This file is becoming quite long.

**Comment by yjbanov** [comment]
LGTM

File: bin/build.dart (hunk 4 above)
Comment (devoncarew) [inline bin/build.dart]: I'll split this out into lib/src/analysis.dart.

File: bin/build.dart (hunk 5 above)
Comment (devoncarew) [inline bin/build.dart]: The `result` value is ignored right now - this is really just a hook to let the benchmark copy its json file into the data directory. I'll add the historical results, so a specific benchmark can decide what it wants to do with it.

**Comment by devoncarew** [comment]
refactoring and merging...

**Comment by devoncarew** [comment]
@yjbanov, can you patch this into the dashboard box? Thanks!

**Pull request merged**